### PR TITLE
Various refactoring around telemetry

### PR DESCRIPTION
**Pull request by Evangelink**

### CODE REVIEW

File: src/Integration/Telemetry/TelemetryManager.cs [hunk 1]
```diff
@@ -67,92 +62,86 @@ public bool IsAnonymousDataShared
...
-            string path = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "msenv.dll");
-            visualStudioVersion = File.Exists(path)
-                ? FileVersionInfo.GetVersionInfo(path).ProductVersion
-                : string.Empty;
+            this.knownUIContexts = knownUIContexts;
 
             if (this.telemetryRepository.Data.InstallationDate == DateTime.MinValue)
             {
-                this.telemetryRepository.Data.InstallationDate = DateTime.Now;
+                this.telemetryRepository.Data.InstallationDate = DateTime.Now; // TODO: Use some mockable clock
```
Comment (SonarTech) [inline src/Integration/Telemetry/TelemetryManager.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: src/Integration/Telemetry/TelemetryManager.cs [hunk 2]
```diff
@@ -67,92 +58,86 @@ public bool IsAnonymousDataShared
...
                 this.telemetryRepository.Save();
             }
 
-            this.tryUploadDataTimer = new Timer { Interval = MillisecondsToWaitBetweenUpload, AutoReset = true };
-            this.firstCallDelayer = new Timer(DelayBeforeFirstUpload) { AutoReset = false };
+            this.firstCallDelayer = timerFactory.Create();
```
Comment (SonarTech) [inline src/Integration/Telemetry/TelemetryManager.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration/Telemetry/TelemetryManager.cs [hunk 3]
```diff
@@ -67,92 +58,86 @@ public bool IsAnonymousDataShared
...
-            string path = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "msenv.dll");
-            visualStudioVersion = File.Exists(path)
-                ? FileVersionInfo.GetVersionInfo(path).ProductVersion
-                : string.Empty;
+            this.knownUIContexts = knownUIContexts;
 
             if (this.telemetryRepository.Data.InstallationDate == DateTime.MinValue)
             {
-                this.telemetryRepository.Data.InstallationDate = DateTime.Now;
+                this.telemetryRepository.Data.InstallationDate = DateTime.Now; // TODO: Use some mockable clock
```
Comment (SonarTech) [inline src/Integration/Telemetry/TelemetryManager.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

**Comment by valhristov** [comment]
Extracted the timer and date comparison logic into TelemetryTimer (could have created more classes). I am not sure how to do end-to-end testing... You should review all commits together as my last two are contradicting.

File: src/Integration/Helpers/Clock.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+
+namespace SonarLint.VisualStudio.Integration
+{
+    public class Clock : IClock
+    {
+        public DateTime Now => DateTime.Now;
```
Comment (Evangelink) [inline src/Integration/Helpers/Clock.cs]: Based on talk we had, shall we not use `UtcNow` or even better `DateTimeOffset`?

File: src/Integration/Telemetry/TelemetryTimer.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
+﻿using System;
```
Comment (Evangelink) [inline src/Integration/Telemetry/TelemetryTimer.cs]: Please add the header.

File: src/Integration/Telemetry/TelemetryTimer.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+using System.Timers;
+
+namespace SonarLint.VisualStudio.Integration
+{
+    public interface ITelemetryTimer : IDisposable
```
Comment (Evangelink) [inline src/Integration/Telemetry/TelemetryTimer.cs]: Please move the `IDisposable` from the interface to the class to follow the pattern of other classes.

Comment (Evangelink) [reply src/Integration/Telemetry/TelemetryTimer.cs]: Please also extract the interface into another file.

File: src/Integration/Telemetry/TelemetryTimer.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    {
+        event EventHandler<TelemetryTimerEventArgs> Elapsed;
+
+        void Start();
+        void Stop();
+    }
+
+    public sealed class TelemetryTimerEventArgs : EventArgs
```
Comment (Evangelink) [inline src/Integration/Telemetry/TelemetryTimer.cs]: Please extract into another class.

File: src/Integration/Telemetry/TelemetryTimer.cs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    {
+        public TelemetryTimerEventArgs(DateTime signalTime)
+        {
+            SignalTime = signalTime;
+        }
+
+        public DateTime SignalTime { get; }
```
Comment (Evangelink) [inline src/Integration/Telemetry/TelemetryTimer.cs]: Yeah immutable data!

File: src/Integration/Telemetry/TelemetryTimer.cs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+                timer.Elapsed -= TryRaiseEvent;
+                timer.Stop();
+            }
+        }
+
+        private void TryRaiseEvent(object sender, ElapsedEventArgs e)
+        {
+            // After the first event we change the interval to 6h
+            timer.Interval = MillisecondsBetweenUploads;
```
Comment (Evangelink) [inline src/Integration/Telemetry/TelemetryTimer.cs]: Weird at first but this actually makes more sense.

File: src/Integration/Telemetry/TelemetryTimer.cs [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+        private readonly ITelemetryDataRepository telemetryRepository;
+        private readonly IClock clock;
+        private readonly ITimer timer;
+
+        private bool isStarted;
+
+        public event EventHandler<TelemetryTimerEventArgs> Elapsed;
+
+        public TelemetryTimer(ITelemetryDataRepository telemetryRepository, IClock clock, ITimer timer)
```
Comment (Evangelink) [inline src/Integration/Telemetry/TelemetryTimer.cs]: Feel free to ignore but I would go for `ITimerFactory` here.

File: src/Integration/Telemetry/TelemetryManager.cs [hunk 11]
```diff
@@ -28,32 +27,19 @@
...
 {
-    [Export(typeof(ITelemetryManager))]
-    [PartCreationPolicy(CreationPolicy.Shared)] // MEF Singleton
-    public sealed class TelemetryManager : ITelemetryManager
+    public sealed class TelemetryManager : ITelemetryManager, IDisposable
     {
-        private const int MinHoursBetweenUpload = 5;
-        private const double MillisecondsToWaitBetweenUpload = 1000 * 60 * 60 * 5; // 5 hours
-        private const double DelayBeforeFirstUpload = 1000 * 60 * 5; // 5 minutes
+        internal static readonly string SonarLintVersion = GetSonarLintVersion();
```
Comment (Evangelink) [inline src/Integration/Telemetry/TelemetryManager.cs]: We might want to extract these 2 methods into some helper.

File: src/Integration/Telemetry/TelemetryManager.cs [hunk 12]
```diff
@@ -28,32 +27,19 @@
 
 namespace SonarLint.VisualStudio.Integration
 {
-    [Export(typeof(ITelemetryManager))]
-    [PartCreationPolicy(CreationPolicy.Shared)] // MEF Singleton
-    public sealed class TelemetryManager : ITelemetryManager
+    public sealed class TelemetryManager : ITelemetryManager, IDisposable
```
Comment (Evangelink) [inline src/Integration/Telemetry/TelemetryManager.cs]: Not worth to do it now but this class could be split into a couple of smaller ones (with smaller responsabilities).

File: src/Integration/Helpers/Clock.cs (hunk 4 above)
Comment (valhristov) [reply src/Integration/Helpers/Clock.cs]: Definitely I think there should be additional properties - UtcNow, Today, etc.

File: src/Integration/Telemetry/TelemetryManager.cs (hunk 12 above)
Comment (valhristov) [reply src/Integration/Telemetry/TelemetryManager.cs]: Perhaps some other day :)

File: src/Integration/Telemetry/TelemetryManager.cs [hunk 13]
```diff
@@ -67,132 +46,116 @@ public bool IsAnonymousDataShared
...
+            knownUIContexts.SolutionBuildingContextChanged += this.OnAnalysisRun;
+            knownUIContexts.SolutionExistsAndFullyLoadedContextChanged += this.OnAnalysisRun;
         }
 
-        public async void OptOut()
+        private TelemetryPayload GetPayload(TelemetryData telemetryData)
         {
-            this.telemetryRepository.Data.IsAnonymousDataShared = false;
-            this.telemetryRepository.Save();
+            var numberOfDaysSinceInstallation = (long)DateTime.Now.DaysPassedSince(telemetryData.InstallationDate);
```
Comment (SonarTech) [inline src/Integration/Telemetry/TelemetryManager.cs]: MINOR Remove this unnecessary cast to 'long'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1905)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Integration/Helpers/IKnownUIContexts.cs [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using Microsoft.VisualStudio.Shell;
+
+namespace SonarLint.VisualStudio.Integration
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Integration/Helpers/IKnownUIContexts.cs]: I think this is better to keep it as we said in the previous review. This is the name that was chosen by Microsoft so it is better to keep it.

File: src/Integration/Telemetry/TelemetryHelper.cs [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+
+namespace SonarLint.VisualStudio.Integration
+{
+    public static class TelemetryHelper
```
[the remark this answers is not in the document]
Comment (valhristov) [reply src/Integration/Telemetry/TelemetryHelper.cs]: We already have that :) But it is doing other things...

File: src/Integration/Helpers/IKnownUIContexts.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (valhristov) [reply src/Integration/Helpers/IKnownUIContexts.cs]: May be it should be IKnownUIContextEvents... But I don't want to change it right now.

File: src/Integration.UnitTests/Helpers/DateTimeExtensionsTests.cs [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+            today_00h_00m.DaysPassedSince(today_00h_01m).Should().Be(0);
+            today_00h_00m.DaysPassedSince(today_23h_59m).Should().Be(0);
+            today_00h_00m.DaysPassedSince(tomorrow_00h_00m).Should().Be(-1);
+            today_00h_00m.DaysPassedSince(tomorrow_23h_59m).Should().Be(-1);
+        }
+
+        [TestMethod]
+        public void HoursPassedSince()
+        {
+            today_00h_00m.HoursPassedSince(yesterday_23h_59m).Should().Be(0);
```
[the remark this answers is not in the document]
Comment (valhristov) [reply src/Integration.UnitTests/Helpers/DateTimeExtensionsTests.cs]: How did you saw this?! :)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: src/Integration/Helpers/IKnownUIContexts.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (michalb-sonar) [reply src/Integration/Helpers/IKnownUIContexts.cs]: Ah, ok. It's a wrapper around class in VS. In that case, the name is fine. Still confusing, but consistent.

File: src/Integration/Telemetry/TelemetryHelper.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (michalb-sonar) [reply src/Integration/Telemetry/TelemetryHelper.cs]: It could also do that - these are all things related to version so I'd put them together.

**Pull request merged**